### PR TITLE
feat(agentos): #WZ-34091 restrict scheduler execution to off-peak time windows

### PR DESCRIPTION
**Pull request by frederic-delsert-whoz**
**Context**

The scheduler was running continuously 24/7, competing with live user traffic for LLM capacity. The goal is to restrict execution to configurable time windows so LLM resources are preserved for live users during business hours.

**What was implemented**

Time-window guard on both phases of the scheduler, implemented via environment variable — without relying on the existing start/pause mechanism (`SchedulerEndpoint`). The window is evaluated dynamically from config + current time on every tick, including after a restart (no persisted state). When no windows are configured, the scheduler runs continuously (existing behaviour preserved).

 **New `ExecutionWindowService`** (`@Service`): receives `SchedulerProperties` and a `Clock` (default `Clock.systemUTC()`) injected by Spring, parses each `DAYOFWEEK HH:mm UTC` boundary into a weekly minute offset at startup (`init` block), groups them pairwise into open/close execution windows, and evaluates whether the current `Instant` falls within any of them via `isWithinExecutionWindow()`. The `Clock` injection makes the service fully testable without mocking. Handles wrap-around windows (e.g. `FRIDAY 22:00 → MONDAY 05:00`). Validates parity, boundary syntax, zero-length windows, ordering/overlap, and allows at most one wrap-around window. Fails open on any invalid config so a misconfiguration never silently halts all scheduled executions.

**New config property** `agentos.prompt.scheduler.windows` (`AGENTOS_PROMPT_SCHEDULER_WINDOWS`), bound as a `List<String>` — Spring splits the comma-separated value, so the whole schedule stays in a single environment variable:

```
MONDAY 22:00,FRIDAY 05:00,FRIDAY 22:00,MONDAY 05:00
```

Boundaries alternate open/close: the example above defines two execution windows — `Mon 22:00 → Fri 05:00` and `Fri 22:00 → Mon 05:00`. When unset (empty list), the scheduler runs continuously.

The open/close role is carried by position in the list rather than by a structured type — a deliberate trade-off to keep the configuration in a single environment variable (structured YAML binding would require one env var per boundary in production).

**Behaviour:**

- **Phase A** (`tickClaim`): outside the window, the tick is a no-op — due prompts accumulate in DB and are processed at the next window open, including after a restart
- **Phase B** (DB poller in `ScheduledPromptExecutor`): the poller skips `claimBatch()` and delays when `consumePaused || !isWithinExecutionWindow()`. Workers drain whatever is already buffered in the channel (at most `channelCapacity` UserRuns, default 50), then block on `receive()` — bounded and deliberate (no lease expiry risk, no duplicate execution)

### CODE REVIEW

**Comment by frederic-delsert-whoz** [comment]
## Code Review — PR #1282: `feat: wz-34091 restrict scheduler execution to off-peak time windows`

The feature is well-scoped: a clean `ExecutionWindowService` utility, a minimal touch on `SchedulerScanner`, good test coverage across boundary conditions, wrap-around, and invalid configs. Four issues need addressing before merge.

---

### 1. Code-style violation — early `return` in `tickClaim()` (`SchedulerScanner.kt`)

```kotlin
when {
    claimPaused.get() -> logger.debug { "... PAUSED ..." }
    !checkExecutionWindow() -> return          // ← early return mid-method
    else -> processClaim()
}
```

The AgentOS code-style rule explicitly states: "Avoid early returns in the middle of a method". The `when` block currently mixes three different exit semantics (log + fall-through, explicit `return`, and `processClaim()`). Restructure so there is a single exit point at the end of the method.

---

### 2. KDoc inconsistency — `parseAndValidate` (`ExecutionWindowService.kt`, line ~110)

```kotlin
 * Returns an empty list on parse failure (fail-open — [isWithinWindow] returns true).
```

The implementation actually returns `null` on parse failure, not an empty list. The `null` return is the correct choice (it is what `isWithinWindow` checks against), but the KDoc is wrong and will mislead future maintainers.

---

### 3. Ordering validation gap for overlapping wrap-around windows (`ExecutionWindowService.kt`, `parseAndValidate`)

The overlap check is:
```kotlin
if (w.open.minuteOfWeek <= prev.close.minuteOfWeek) { errors += ... }
```

This compares raw `minuteOfWeek` integers. It correctly rejects non-wrap-around overlaps. However, two wrap-around windows (both with `close < open`, e.g. two `SUNDAY xx:00 → MONDAY yy:00` entries) could overlap each other without being detected, because the raw `close` value of the first is small and the `open` of the second is large — the comparison passes even though the windows overlap in the Sunday→Monday zone. Minor edge case in practice, but a correctness gap in the validator.

---

### 4. Misleading UTC label in log message (`SchedulerScanner.kt`, `checkExecutionWindow`)

```kotlin
logger.info { "... (${now.toLocalTime()} UTC)" }
```

`now` is constructed from the injected `clock` whose zone is not guaranteed to be UTC. In tests `clock` is fixed to a UTC instant but nothing enforces this in production. If the clock zone is ever non-UTC, `toLocalTime()` returns the wall-clock time in that zone while the label says "UTC" — misleading. The time should be converted to UTC before extracting the local time for the log string.

File: agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+        parsedWindows = parseAndValidate(windows)
+    }
+
+    /**
+     * Returns `true` when [now] falls within any configured execution window.
+     *
+     * Returns `true` unconditionally when no windows are configured (always-open behaviour)
+     * or when configuration parsing failed (fail-open).
+     */
+    fun isWithinWindow(now: ZonedDateTime = ZonedDateTime.now(ZoneOffset.UTC)): Boolean {
```
Comment (vincent-couturier-whoz) [inline agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt]: question: Why do you use ZonedDateTime instead of Instant which is by design a UTC datetime ?

File: agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        }
+
+        // Validate ordering: each open must be strictly after the previous close.
+        // Skipped when a wrap-around pair was already reported — raw offset comparison is
+        // meaningless in that case and would emit a confusing secondary error.
+        if (wrapAroundWindows.size <= 1) {
+            for (i in 1 until windows.size) {
+                val w = windows[i]
+                val prev = windows[i - 1]
+                if (w.open == w.close || prev.open == prev.close) continue // already reported above
```
Comment (vincent-couturier-whoz) [inline agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt]: suggestion: use collection operator to have a more functional programming style and so avoid bug due to wrong index or wrong continue effect.

**Review by vincent-couturier-whoz — DISMISSED** [review]

File: agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt (hunk 1 above)
Comment (frederic-delsert-whoz) [reply agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt]: done at commit: ea707d391163b5e9965de6b080a24f79a8b62088

File: agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt (hunk 2 above)
Comment (frederic-delsert-whoz) [reply agentos/agentos-service/src/main/kotlin/io/whozoss/agentos/scheduledPrompt/ExecutionWindowService.kt]: done at commit: 94ea36bdddcbf7f74030beb555e014889e65c3aa

**Review by vincent-couturier-whoz — APPROVED** [review]

**Review by nicolas-remy-whoz — APPROVED** [review]
